### PR TITLE
Improve shairport-sync integration

### PR DESCRIPTION
**Pull request by dgonano**
This PR allows music playing through shairport to be,
* muted,
* reduced by x%, 
* do nothing,
 
while Alexa is listening, processing and speaking. 
It reverses the set command to return to previous volume once Alexa interaction finishes. 

Main.py
* removed redundant while loop
* added start & stop interaction commands if shairport is enabled 

setup.sh
* Add section to correctly configure audio (once decision is made on to keep the config file or use a new one)
* update config.yaml with enabled = true/false
* update asound.conf so ALSA has the new soft device
* update shariport-sync.conf so it points audio to the new soft device for control 

Uninstall has been tested to. it comments out the changes made to shairport-sync so that shairport sends audio to the "default" device and is not controlled by AlexaPi

### CODE REVIEW

**Comment by renekliment** [comment]
Hey, thank you for this! 
I'll take a look later, since this is a new feature and we'll therefore be pushing it to v1.4.

**Comment by dgonano** [comment]
Yeah thats okay. I suggest we set up proper branches in the repository through so that things like this can be merged into a development branch. To do this properly we need to agree on a workflow. 
See: http://blog.endpoint.com/2014/05/git-workflows-that-work.html 
I suggest `gitflow` for discrete releases and `github flow` for continuous releases.  

By what your saying i'm assuming you'd like gitflow with proper releases (see main picture at the top of that link and also description down the bottom). Assuming this, at a minimum we need `Master` & `Development` branches. 
At a time like this though we should also have `v1.3release` so we can tidy up v1.3 ready to release in an isolated area. 
* Bugfix\hotfix would go straight into `Master` and be merged into `development` (and possibly be merged into `v1.3release` depending on how it affects it).
* Get everything ready for release of 1.3 in its own branch. Merge into master only when ready to release immediately. Would then update Development branch from master to ensure people are developing on the latest.  
* This stuff that will probably go in go into 1.4 should go straight into development branch

**Comment by renekliment** [comment]
Yes, I'm planning something like this. I was thinking doing this after _v1.4_, but if you feel like it's something we should do now, let's do it after _v1.3_.

**Comment by dgonano** [comment]
We need a Development branch now because we have concurrent development. Code shouldn't be left sitting in personal repo's because it'll become out of date. Everyone already working on things for 1.4 should be submitting PRs to the dev branch. Even a feature branch for big ones like API upgrades.

Anyway i reckon put it in a dev branch now and people will start using, testing (just by using) etc It's not too important because it's going to get proper testing coming up to the 1.4 release. Plus others working on 1.4 stuff can now submit PRs too. 

Use master now as release branch for 1.3. Tag it when its's ready and merge into development so dev is up to date. 

just my 2 cents.

**Comment by renekliment** [comment]
Alright, let's do it after _v1.3_.

**Comment by renekliment** [comment]
So I've read the article and if I'm not wrong we currently do _github flow_.

> Code shouldn't be left sitting in personal repo's because it'll become out of date.

This doesn't really happen apart from this 2 week period before a release (or something we call release, but it's more of a just tag).

**Comment by dgonano** [comment]
No, we are not following github flow for gitflow at the moment. 

Githubflow:
* You don't do preparation for major versioning releases. Instead you do continuous releases whenever a feature is finished (no matter the order). A major release may be declared when a major feature is merged into master but this is just semantics. 
* There is no such thing as a hold on the repo to prepare for a release. Its either tested and released or not in master, that simple. As soon as you do this you break the system because you've put a hold on the ONLY branch (master).
* Document updates must be done with every PR because **every PR is a deployable release**.

gitflow: 
* Traditionally used when doing discreet releases, e.g. 1.2, 1.3, 2.0 etc. 
* Master is the stable branch that is very controlled, versions are released from here
* Development branch is where development PRs get put. 
* When preparing for a release (like we are now) one would branch from development into a special release branch e.g. `v1.3 release` so it can be tightly controlled, prepared and heavily tested for release. This is required because it's about to be merged into master and released. This is effectively what you have done at the moment except instead of locking up the special branch, the whole repository has been locked. 
* Development would continue while this release prep is happening. After the release where `v1.3 release` is merged into master and released, master would be merged into development to ensure development is up to date. 
* Bugfix/hotfix can go straight into master and released as they are urgent bug fix's

Hope that makes sense :)

**Comment by renekliment** [comment]
I see, it's the hold up that makes the difference then.
As I've mentioned somewhere else, the docs should be updated with every PR now (if relevant). It's been a mistake / laziness so far that it hasn't been updated. (Mostly me being lazy not specifying it.)

The reasons I want to do something as _releases_:
1. I think it's not the worst idea to pause once in a while for a week or two, to make sure that `master` _really is_ bug-free.
2. I like doing the milestones (releases) for planning stuff, so people (and I) have idea what's the priority for the near future and what should come before what. This is not static though - if someone else does something sooner, it goes into the nearest _release_.

Maybe the real question is what do we want instead of what do we do now. It's not my decision only. I have a vision and I am kind of the self-appointed _project leader_ (or better a coordinator), but everything about AlexaPi should stay open as it is an open project after all.

@alexa-pi/coreteam What do **you** want for AlexaPi? (github flow vs. gitflow) We can put it up to a vote.

**Comment by illperipherals** [comment]
I personally like the idea of milestone releases. This is not to say we need to adopt a waterfall methodology, but we don't have to "release" at each build.

My .02

**Comment by dgonano** [comment]
Hey all, 
Of course I agree that the docs should be updated and I'm probably guilty of doing it well enough myself :(
I have no issue with stopping and doing a proper release, I think it's probably the better way to do it. 

@renekliment I didn't mean to disrespect how you have been handling this project at all, sorry if it's come across like that. 
I was just trying to say that as we get more developers it would likely be beneficial to have a development bach (and generally following gitflow). 
I'm happy for this to be added once 1.3 is done though :)

**Comment by renekliment** [comment]
@dgonano Nope, it's fine. Please do speak up when you feel there is a better way to do something. I just wanted to express that it's not my decision only and everything is open for discussion - it's just that most of the time no one says anything so I do it my way :smile: 

> I'm happy for this to be added once 1.3 is done though :)

You mean the PR?

**Comment by dgonano** [comment]
> You mean the PR?

A development branch, that the PR (and all development PRs) can be merged into. 
Only bug fix's going for immediate release (eg. 1.3.1) should go into master.

File: src/config.template.yaml [hunk 1]
```diff
@@ -26,6 +26,17 @@ sound:
   media_output_device:
   media_default_volume:
 
+shairport:
+  # Is shairport installed & running
+  enabled: false
+
+  # Actions to perfomr on shairport-sync music while listening/speaking
```
Comment (renekliment) [inline src/config.template.yaml]: typo _perfomr_

File: src/scripts/setup.sh [hunk 2]
```diff
@@ -188,6 +194,45 @@ case ${config_action} in
 
 esac
 
+# Update shairport config now we know the config option
```
Comment (renekliment) [inline src/scripts/setup.sh]: Move all this to _inc/common.sh_, or better, create an _inc/shairport-sync.sh_ and put it there as a function like `shairport-sync_setup` along with the variables you have defined. Also, when there is this file, it's a good idea to move the `function install_shairport-sync_from_source` from _inc/common.sh_ into the new file. Don't forget to deal with `${TMP_DIR}` and to include this new file in _setup.sh_.

The setup script is kind of like a wizard and all the magic happens in the _inc/_ files.

File: src/scripts/setup.sh [hunk 3]
```diff
@@ -11,6 +11,12 @@ CONFIG_FILENAME="config.yaml"
 CONFIG_FILE_SYSTEM="${CONFIG_SYSTEM_DIRECTORY}/${CONFIG_FILENAME}"
 CONFIG_FILE_LOCAL="./${CONFIG_FILENAME}"
 
+ALSA_CONFIG_FILE="/etc/asound.conf"
+ALSA_CONFIG_FILE_TEMPLATE="${SCRIPT_DIRECTORY}/asound.template.conf"
+SHAIRPORT_CONFIG_FILE="/usr/local/etc/shairport-sync.conf"
+SHAIRPORT_ALSA_DEVICE="shairportdev"
+SHAIRPRT_ALSA_CONTROL="shairportvolume"
```
Comment (renekliment) [inline src/scripts/setup.sh]: typo _SHAIRPRT_
Also, it doesn't seem to be used anywhere.

File: src/scripts/asound.template.conf [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+	card 0
+}
+
+ctl.!default {
+	type hw
+	card 0
+}
+
+
+pcm.shairportdev {
```
Comment (renekliment) [inline src/scripts/asound.template.conf]: Can you maybe put this stuff into an extra config file so we don't have to rewrite the main ALSA config? I'm not sure what's the best approach here.

**Comment by renekliment** [comment]
Also, please do the PR against the `dev` branch. Thank you.

**Comment by renekliment** [comment]
... aaand update the CHANGELOG.md please. I should probably create a PR template so we don't forget things like this.

**Comment by dgonano** [comment]
Yup good idea. 
I've been away but back now so I'll work through all the PRs soon.

**Comment by renekliment** [comment]
@dgonano Dope! In the words of SNL: "We've got a great show tonight!". There are some sweet things waiting for review, along with the almost-finished logging branch from @lechat.
I'd like to release _v1.4_ rather sooner than later - I've already pushed some things like refactoring the AVS code into the 1.5 milestone so we can roll out _v1.4_ soon. I think snowboy and covering the functionality of the majority of the works is a big enough _feature_ / milestone and we should release that as soon as possible. Let's honor the "release early, release" often scheme.

**Comment by renekliment** [comment]
So I gave this another thought and I think we should not hardwire shairport in the python code. Instead, let's make config options for the commands that happen _pre-_ and _post-_ request. People also might want to mute mpd or another player.

**Comment by dgonano** [comment]
@renekliment I just rebased and resolved this PR to keep it up to date but I don't think theres much point finishing this until the shellcheck and triggering PRs are merged as they are pretty big changes (which this has to work in with)

**Comment by dgonano** [comment]
Even better, should we setup pre and post bash scripts to run? That way people can put what ever commands, etc they like in them. 
eg, `pre-interaction.sh` and `post-interaction.sh`
And add them in `/etc/opt/AlexaPi/` with the config so they are persistent over clean installs.

**Comment by renekliment** [comment]
@dgonano Alright, makes sense.

Let's just not bother and do configurable commands in the config file. People would often use single line commands anyway - like here or for other players like `mpc volume 20` and if they do need whole scripts, they can easily set the script path as the command.

**Comment by dgonano** [comment]
Closing, replaced with PR #125

**Pull request closed without merging**